allennlp/data/semparse/worlds/nlvr_world.py

```diff
@@ -0,0 +1,436 @@
+                target_attribute: AttributeType,
+                comparison_op: Callable[[AttributeType, AttributeType], bool]) -> Set[EntityType]:
+        returned_set = []
+        for entity in set_to_filter:
+            if comparison_op(attribute_function(entity), target_attribute):
+                returned_set.append(entity)
+        return set(returned_set)
+
+    ## Box filtering functions
+    @classmethod
+    def filter_equals(cls,
+                      set_to_filter: Set[Box],
+                      attribute_function: Callable[[Box], AttributeType],
+                      target_attribute: AttributeType) -> Set[Box]:
+        return cls._filter(set_to_filter, attribute_function, target_attribute, operator.eq)
+
+    @classmethod
+    def filter_not_equal(cls,
+                         set_to_filter: Set[Box],
+                         attribute_function: Callable[[Box], AttributeType],
+                         target_attribute: AttributeType) -> Set[Box]:
+        return cls._filter(set_to_filter, attribute_function, target_attribute, operator.ne)
+
+    @classmethod
+    def filter_greater_equal(cls,
+                             set_to_filter: Set[Box],
+                             attribute_function: Callable[[Box], AttributeType],
+                             target_attribute: AttributeType) -> Set[Box]:
+        return cls._filter(set_to_filter, attribute_function, target_attribute, operator.ge)
+
+    @classmethod
+    def filter_lesser_equal(cls,
+                            set_to_filter: Set[Box],
+                            attribute_function: Callable[[Box], AttributeType],
+                            target_attribute: AttributeType) -> Set[Box]:
+        return cls._filter(set_to_filter, attribute_function, target_attribute, operator.le)
+
+    ## Object filtering functions
+    @classmethod
+    def black(cls, objects_set: Set[Object]) -> Set[Object]:
+        return cls._filter(objects_set, cls._get_single_color, 'black', operator.eq)
+
+    @classmethod
+    def blue(cls, objects_set: Set[Object]) -> Set[Object]:
+        return cls._filter(objects_set, cls._get_single_color, 'blue', operator.eq)
+
+    @classmethod
+    def yellow(cls, objects_set: Set[Object]) -> Set[Object]:
+        return cls._filter(objects_set, cls._get_single_color, 'yellow', operator.eq)
+
+    @classmethod
+    def circle(cls, objects_set: Set[Object]) -> Set[Object]:
+        return cls._filter(objects_set, cls._get_single_shape, 'circle', operator.eq)
+
+    @classmethod
+    def square(cls, objects_set: Set[Object]) -> Set[Object]:
+        return cls._filter(objects_set, cls._get_single_shape, 'square', operator.eq)
+
+    @classmethod
+    def triangle(cls, objects_set: Set[Object]) -> Set[Object]:
+        return cls._filter(objects_set, cls._get_single_shape, 'triangle', operator.eq)
+
+    @classmethod
+    def same_color(cls, objects_set: Set[Object]) -> Set[Object]:
```

Comment:
This name is confusing for what the function does.  I would have expected this function to return a boolean saying whether all objects in the set have the same color.

Well, I guess `triangle` doesn't return a boolean, it just filters the set to things that are triangles.  Here you're filtering the set to things that all have the same color, without specifying a priori what that color is.  If you explain it that way clearly in the docstring, I think the name is fine.

Reply:
I think the docstring does convey what it's actually doing. Not sure how it could be clearer.

Reply:
Yes, the docstring does a good job saying what the function _does_, it just doesn't motivate _why_, which is what you need to understand the name. Adding a sentence like this would probably be enough: "Just as `triangle` filters the set to objects that are triangles, this method filters the set to only contain objects of the same color." The semantics of the function are just a little odd given the name, so some extra clarification in the docstring would be nice.